src/library/scala/collection/SeqLike.scala

```diff
@@ -504,15 +504,18 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
    *  @return  A new $coll which contains the first occurrence of every element of this $coll.
    */
   def distinct: Repr = {
-    val b = newBuilder
-    val seen = mutable.HashSet[A]()
-    for (x <- this) {
-      if (!seen(x)) {
-        b += x
-        seen += x
+    if (lengthCompare(1) <= 0) repr
```

Comment:
I just realized (after discussions about https://github.com/scala/bug/issues/10415#issuecomment-368047800) that this implementation applies to mutable collections, too. We can't simply return `repr` there, it has to be a copy. There is no `distict` implementation specifically for for immutable collection so you'll have to handle this case here (because we can't introduce a new override).

```
val isImmutable = this.isInstanceOf[immutable.Seq[_]]
if(isImmutable && lengthCompare(1) <= 0) repr
else {
  ...
  if(different || !isImmutable) b.result() else repr
}
```

Reply:
so we should also have a follow-on change targeted at 2.13 only to introduce this in immutable.Seq